crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,455 @@
+use std::cmp::Ordering;
+
+use ruff_diagnostics::{Applicability, Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_ast::whitespace::indentation;
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::{lexer, Mode, Tok};
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use crate::checkers::ast::Checker;
+
+use itertools::Itertools;
+
+/// ## What it does
+/// Checks for `__all__` definitions that are not alphabetically sorted.
+///
+/// ## Why is this bad?
+/// Consistency is good. Use a common convention for `__all__` to make your
+/// code more readable and idiomatic.
+///
+/// ## Example
+/// ```python
+/// import sys
+///
+/// __all__ = [
+///     "b",
+///     "c",
+///     "a",
+/// ]
+///
+/// if sys.platform == "win32":
+///     __all__ += ["z", "y"]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import sys
+///
+/// __all__ = [
+///     "a",
+///     "b",
+///     "c",
+/// ]
+///
+/// if sys.platform == "win32":
+///     __all__ += ["y", "z"]
+/// ```
+///
+/// ## Fix safety
+/// This rule's fix should be safe for single-line `__all__` definitions
+/// and for multiline `__all__` definitions without comments.
+/// For multiline `__all__` definitions that include comments,
+/// the fix is marked as unsafe, as it can be hard to tell where the comments
+/// should be moved to when sorting the contents of `__all__`.
+#[violation]
+pub struct UnsortedDunderAll;
+
+impl Violation for UnsortedDunderAll {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__all__` is not alphabetically sorted")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Sort `__all__` alphabetically".to_string())
+    }
+}
+
+/// RUF022
+pub(crate) fn sort_dunder_all(checker: &mut Checker, stmt: &ast::Stmt) {
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    // We're only interested in `__all__ = ...` and `__all__ += ...`
+    let (target, original_value) = match stmt {
+        ast::Stmt::Assign(ast::StmtAssign { value, targets, .. }) => match targets.as_slice() {
+            [ast::Expr::Name(ast::ExprName { id, .. })] => (id, value.as_ref()),
```

Comment:
I always use `.as_ref()` for this, FWIW. I might be doing it wrong then :)

Reply:
Yeah that's the evil thing about `AsRef::as_ref()`. It works, and then you upgrade Rust, and maybe it breaks. We haven't gotten one in a while, but there have been lots of reports of breakage in the past because std added a trait impl somewhere.

(This is an instance of a more general problem where adding trait impls can cause inference to fail where it previously succeeded. `AsRef` just happens to be one particular instance of it.)

With all that said, the worst thing that happens is our code will stop compiling when a new version of Rust is released. And a fix is (guaranteed actually) to be simple, so long as you diagnose it correctly. You can never get something wrong. But... these things stand out to me because I've been bitten by it myself and users reporting breakage to me have been bitten by it.

Reply:
I had a bit of trouble locally getting it to work without `.as_ref()` :(

@burntsushi, would you mind pasting a diff of what you mean here? Or feel free to push straight to my branch if you prefer, whichever's easiest :)

Reply:
Yeah! So it looks like there are only two `.as_ref()` calls left here. This diff is the simplest one that removes them:

```
[andrew@duff ruff]$ g diff
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
index d42249f38..227454cea 100644
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
@@ -101,7 +101,7 @@ pub(crate) fn sort_dunder_all_aug_assign(
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = &**target else {
         return;
     };
     sort_dunder_all(checker, id, value, parent);
@@ -122,7 +122,7 @@ pub(crate) fn sort_dunder_all_ann_assign(
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = &**target else {
         return;
     };
     sort_dunder_all(checker, id, val, parent);
```

The key here is noticing that `target` has type `&Box<Expr>`. Because you can't pattern match through a `Box` (a limitation that everyone would like to lift, but is I guess difficult to do so), and since you want to borrow from the AST, you want a `&Expr` here. To go from a `&Box<Expr>`, you can dereference twice to get an `Expr` (temporarily), and then add a `&` to get a `&Expr`.

Now for me personally, this is not how I would write this. The only reason why it works is something called [match ergonomics](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html) that essentially tries to remove the need to explicitly mark whether a pattern match is borrowed or not. I think match ergonomics makes code like this less clear because it's harder to tell at a glance what `target` is just from reading the source code.

This is how I would write it:

```
[andrew@duff ruff]$ g diff
diff --git a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
index d42249f38..ed3ee0340 100644
--- a/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs
@@ -93,15 +93,15 @@ pub(crate) fn sort_dunder_all_aug_assign(
     parent: &ast::Stmt,
 ) {
     let ast::StmtAugAssign {
-        value,
-        target,
+        ref value,
+        ref target,
         op: ast::Operator::Add,
         ..
-    } = node
+    } = *node
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { ref id, .. }) = **target else {
         return;
     };
     sort_dunder_all(checker, id, value, parent);
@@ -115,14 +115,14 @@ pub(crate) fn sort_dunder_all_ann_assign(
     parent: &ast::Stmt,
 ) {
     let ast::StmtAnnAssign {
-        target,
-        value: Some(val),
+        ref target,
+        value: Some(ref val),
         ..
     } = node
     else {
         return;
     };
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+    let ast::Expr::Name(ast::ExprName { ref id, .. }) = **target else {
         return;
     };
     sort_dunder_all(checker, id, val, parent);
```

Now you still need the `**target`, but the fact that you need it is now (maybe) a little clearer. That match on `ref target` tells you that `target` _has_ to be a `&SOMETHING`.

Reply:
Thanks so much <3. I applied your suggestion in https://github.com/astral-sh/ruff/pull/9474/commits/229fec17edf6aa1b1ed2c2a365b9e81e740908db.

I _think_ I understand the solution, but will try to read up a bit more this weekend to check I fully understand what's going on here :)

Reply:
Yeah it will take some getting used to. There's a reason why match ergonomics exists even if I'm no fan. :-)